File: src/PageBuilder.js

const path = require('path');
const ejs = require('ejs');
const fs = require('fs');

const Compressor = require('./AssetCompressor.js');
const Utils = require('./Utils.js');

class PageBuilder
{
	constructor(isRelease, compress, pathToRoot, siteConfig, projectConfig, iconsConfig)
	{
		this.IsRelease = isRelease;
		this.Compress = compress;
		this.PathToRoot = pathToRoot;

		this.SiteConfig = siteConfig;
		this.ProjectConfig = projectConfig;
		this.IconsConfig = iconsConfig;

	}

	BuildPage(pagePath, config)
	{
		let sourcePath = path.join(this.PathToRoot, this.SiteConfig.Raw_ViewsFolder, pagePath);
		let outputPath = path.join(this.PathToRoot, this.SiteConfig.Output_ViewsFolder, pagePath);

		let sourceFilePath = sourcePath + ".ejs"
		let outputFilePath = outputPath + ".html"


		// check is file
		let stat = fs.statSync(sourceFilePath);
		if (!stat.isFile())
		{
			console.error("Cannot build folder: " + sourceFilePath);
			return;
		}

		Utils.TryMakeDir(path.dirname(outputFilePath));

		let pageName = path.basename(outputFilePath, ".html");
		let pageParent = path.dirname(pagePath);

		// count how many levels up the path is
		let count = pagePath.split("/").length - 1;
		let pathToRoot = "../".repeat(count);


		config["PageData"] = {
			IsRelease: this.IsRelease,
			PageName: pageName,
			PageParent: pageParent,
			PathToRoot: pathToRoot,
			SiteConfig: this.SiteConfig,
			Projects: this.ProjectConfig,
			Icons: this.IconsConfig,
		}

		this.RenderFile(sourceFilePath, outputFilePath, config);
	}

	RenderFile(sourceFile, outputFile, config)
	{
		console.log(Utils.MakePathRelative(sourceFile), "->", Utils.MakePathRelative(outputFile));

		let compress = this.Compress;

		//render selected file
		ejs.renderFile(sourceFile, config,
			function(renderError, builtHtml) {

			// error handling
			if (renderError)
			{
				console.error(renderError);
				return;
			}


			if (compress)
				builtHtml = Compressor.CompressHtml(builtHtml);

			// write rendered page to output path
			fs.writeFile(outputFile, builtHtml, function (fileError) {

				// error handling
				if (fileError)
				{
					console.error(fileError);
					return;
				}
			});
		});
	}
}


exports.PageBuilder = PageBuilder;

File: tools/pkg/builder/pagebuilder.go
package builder

import (
	"fmt"
	"html/template"
	"os"
	"path/filepath"
	"strings"
	"time"

	"sitebuilder/pkg/config"
	"sitebuilder/pkg/templatefuncs"
)

// PageBuilder builds individual HTML pages from templates
type PageBuilder struct {
	isRelease  bool
	pathToRoot string
	siteConfig config.SiteConfig
	projects   map[string]config.Project
	icons      map[string]config.Icon
	baseTmpl   *template.Template
}

// NewPageBuilder creates a new PageBuilder and loads component templates
func NewPageBuilder(isRelease bool, pathToRoot string, siteConfig config.SiteConfig,
	projects map[string]config.Project, icons map[string]config.Icon) (*PageBuilder, error) {

	pb := &PageBuilder{
		isRelease:  isRelease,
		pathToRoot: pathToRoot,
		siteConfig: siteConfig,
		projects:   projects,
		icons:      icons,
	}

	if err := pb.loadComponentTemplates(); err != nil {
		return nil, err
	}

	return pb, nil
}

func (pb *PageBuilder) loadComponentTemplates() error {
	pb.baseTmpl = template.New("base").Funcs(templatefuncs.FuncMap())

	compDir := filepath.Join(pb.pathToRoot, pb.siteConfig.Raw_ViewsFolder, "Components")
	entries, err := os.ReadDir(compDir)
	if err != nil {
		return fmt.Errorf("reading components dir: %w", err)
	}

	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".html") {
			continue
		}
		filePath := filepath.Join(compDir, e.Name())
		content, err := os.ReadFile(filePath)
		if err != nil {
			return fmt.Errorf("reading component %s: %w", e.Name(), err)
		}
		if _, err := pb.baseTmpl.Parse(string(content)); err != nil {
			return fmt.Errorf("parsing component %s: %w", e.Name(), err)
		}
	}

	return nil
}

// BuildPage renders a single page template to the output directory
func (pb *PageBuilder) BuildPage(pagePath string) error {
	sourcePath := filepath.Join(pb.pathToRoot, pb.siteConfig.Raw_ViewsFolder, pagePath+".html")
	outputPath := filepath.Join(pb.pathToRoot, pb.siteConfig.Output_ViewsFolder, pagePath+".html")

	stat, err := os.Stat(sourcePath)
	if err != nil {
		return fmt.Errorf("stat source %s: %w", sourcePath, err)
	}
	if !stat.Mode().IsRegular() {
		return fmt.Errorf("not a regular file: %s", sourcePath)
	}

	if err := os.MkdirAll(filepath.Dir(outputPath), 0755); err != nil {
		return fmt.Errorf("creating output dir: %w", err)
	}

	pageName := filepath.Base(pagePath)
	pageParent := filepath.Dir(pagePath)

	// Count depth for pathToRoot
	parts := strings.Split(pagePath, "/")
	depth := len(parts) - 1
	pathToRoot := strings.Repeat("../", depth)

	// Get project data for this page
	var projectData config.Project
	for _, proj := range pb.projects {
		if proj.PagePath != nil && *proj.PagePath == pagePath {
			projectData = proj
			break
		}
	}

	pageData := config.PageData{
		IsRelease:   pb.isRelease,
		PageName:    pageName,
		PageParent:  pageParent,
		PathToRoot:  pathToRoot,
		SiteConfig:  pb.siteConfig,
		Projects:    pb.projects,
		Icons:       pb.icons,
		CurrentDate: time.Now().Format("2006-01-02"),
	}

	ctx := config.PageContext{
		PageData:    pageData,
		ProjectData: projectData,
	}

	content, err := os.ReadFile(sourcePath)
	if err != nil {
		return fmt.Errorf("reading page %s: %w", sourcePath, err)
	}

	// Clone the base template and parse the page
	tmpl, err := pb.baseTmpl.Clone()
	if err != nil {
		return fmt.Errorf("cloning template: %w", err)
	}

	tmpl, err = tmpl.New("page").Parse(string(content))
	if err != nil {
		return fmt.Errorf("parsing page template %s: %w", pagePath, err)
	}

	outFile, err := os.Create(outputPath)
	if err != nil {
		return fmt.Errorf("creating output file: %w", err)
	}
	defer outFile.Close()

	fmt.Printf("  %s -> %s\n", sourcePath, outputPath)

	if err := tmpl.ExecuteTemplate(outFile, "page", ctx); err != nil {
		return fmt.Errorf("executing template %s: %w", pagePath, err)
	}

	return nil
}
